Modules/Loadable/Models/Logic/vtkSlicerModelsLogic.cxx

```diff
@@ -438,33 +439,102 @@ void vtkSlicerModelsLogic::TransformModel(vtkMRMLTransformNode *tnode,
 
   vtkAbstractTransform *transform = tnode->GetTransformToParent();
   modelOut->ApplyTransform(transform);
+  modelOut->SetAndObserveTransformNodeID(mtnode == NULL ? NULL : mtnode->GetID());
+
+  return;
+}
+
+//----------------------------------------------------------------------------
+void vtkSlicerModelsLogic::MirrorModel(vtkMRMLModelNode *modelNode,
+                                       vtkMRMLModelNode *modelOut,
+                                       bool mirrorX, bool mirrorY, bool mirrorZ)
+{
+  if (!modelNode || !modelOut )
+    {
+    return;
+    }
+
+  vtkSmartPointer<vtkPolyData> poly = vtkSmartPointer<vtkPolyData>::New();
+  modelOut->SetAndObservePolyData(poly);
+
+  poly->DeepCopy(modelNode->GetPolyData());
+
+  double oldPoint[3];
+  int nPoints = poly->GetNumberOfPoints();
+  vtkSmartPointer<vtkPoints> polyPoints = poly->GetPoints();
+
+  for( int PointId = 0; PointId < nPoints; PointId++ )
+  {
+      polyPoints->GetPoint(PointId, oldPoint);
+      double newPoint[3];
+
+      if (mirrorX)
+          newPoint[0] = oldPoint[0] * (-1);
+      else
+          newPoint[0] = oldPoint[0];
+
+      if (mirrorY)
+          newPoint[1] = oldPoint[1] * (-1);
+      else
+          newPoint[1] = oldPoint[1];
+
+      if (mirrorZ)
+          newPoint[2] = oldPoint[2] * (-1);
+      else
+          newPoint[2] = oldPoint[2];
+
+      polyPoints->SetPoint(PointId, newPoint);
+  }
+
+  poly->SetPoints(polyPoints);
+  vtkSmartPointer<vtkPolyDataNormals> normals = vtkSmartPointer<vtkPolyDataNormals>::New();
+  normals->SetInputData(poly);
+  normals->AutoOrientNormalsOn();
+  normals->ConsistencyOn();
+  normals->Update();
 
-  if (transformNormals)
-    {
-    // fix normals
-    //--- NOTE: This filter recomputes normals for polygons and
-    //--- triangle strips only. Normals are not computed for lines or vertices.
-    //--- Triangle strips are broken up into triangle polygons.
-    //--- Polygons are not automatically re-stripped.
-    vtkNew<vtkPolyDataNormals> normals;
-    normals->SetInputData(poly.GetPointer());
-    //--- NOTE: This assumes a completely closed surface
-    //---(i.e. no boundary edges) and no non-manifold edges.
-    //--- If these constraints do not hold, the AutoOrientNormals
-    //--- is not guaranteed to work.
+  modelOut->SetPolyDataConnection(normals->GetOutputPort());
+
+  //modelOut->SetAndObservePolyData(poly);
```

Comment:
Thanks for the feedback! I have updated the code to use vtkTransformPolyDataFilter and there is still the problem described here http://na-mic.org/Mantis/view.php?id=4297 unless normals are explicitly filtered using the vtkPolyDataNormals. Do you think this is something we should report to the VTK folks?